advanced/AddressableLEDController.java

```diff
@@ -0,0 +1,183 @@
+
+        ledNotifier = new Notifier(() -> {
+            LEDStepHandler(ledStepperFunction);
+        });
+        ledNotifier.startPeriodic(period);
+    }
+
+    /**
+     * This is a uniform way to handle most simple custom functions (like a moving
+     * rainbow) to control the LEDs. Every time this is called, it raises the step
+     * counter by 1.
+     * 
+     * @param ledStepperFunction custom LED controller function <step number,
+     *                           LEDbuffer, LEDbuffer>
+     */
+    public void LEDStepHandler(BiFunction<Integer, AddressableLEDBuffer, AddressableLEDBuffer> ledStepperFunction) {
+        ledBuffer = ledStepperFunction.apply(step, ledBuffer);
+        updateLEDs();
+
+        step++;
+    }
+
+    /**
+     * Same as rainbow() except the rainbow is stretch out over the whole length of
+     * the LED strip.
+     * 
+     * @param stepPeriod number of steps for the pattern to move a whole wavelength
+     * @param step       the step number, each step will generally be 1/50th of a
+     *                   second [given by handler]
+     * @param buffer     the LED buffer, it is modified then returned [given by
+     *                   handler]
+     * @return the modified LED buffer
+     */
+    public AddressableLEDBuffer stretchRainbow(int stepPeriod, int step, AddressableLEDBuffer buffer) {
+        return rainbow(stepPeriod, length, length, maxBrightness, step, buffer);
+    }
+
+    /**
+     * And example of a custom LED step function. It is a moving rainbow has a
+     * custom wavelength and movement "speed."
+     * 
+     * @param stepPeriod number of steps for the pattern to move a whole wavelength
+     * @param wavelength length (in LEDs) of the wave, so number of LEDS between the
+     *                   two red points (or any other color)
+     * @param length     length of LED strip
+     * @param step       the step number, each step will generally be 1/50th of a
+     *                   second [given by handler]
+     * @param buffer     the LED buffer, it is modified then returned [given by
+     *                   handler]
+     * @return the modified LED buffer
+     */
+    public static AddressableLEDBuffer rainbow(int stepPeriod, int wavelength, int length, int value, int step,
+            AddressableLEDBuffer buffer) {
+        int degrees = 180;
+        int saturation = 255;
+
+        for (int i = 0; i < length; i++) {
+            int shiftedI = (i + (step * wavelength / stepPeriod)) % length;
+            int hue = (shiftedI * degrees / wavelength) % degrees;
+
+            buffer.setHSV(i, hue, saturation, value);
+        }
+        return buffer;
+    }
+
+    /**
+     * Sets the entire strip to one color using the RGB notation.
+     * 
+     * @param r red value [0, 255]
+     * @param g blue value [0, 255]
+     * @param b blue value [0, 255]
+     */
+    public void setColorRGB(int r, int g, int b) {
+        System.out.println("asdfasdf");
```

Comment:
i don't think this i supposed to be here - i'm assuming it's a debugging line

Reply:
lmaoooo whoops